libr/core/cmd_print.c

```diff
@@ -4486,6 +4486,111 @@ static void r_core_disasm_table(RCore * core, int l, const char *input) {
 	r_table_free (t);
 }
 
+static void cmd_pxr(RCore *core, int len, int mode, int wordsize, const char *arg) {
+	PJ *pj = NULL;
+	RTable *t = NULL;
+	if (mode == ',') {
+		t = r_table_new ();
+		RTableColumnType *n = r_table_type ("number");
+		RTableColumnType *s = r_table_type ("string");
+		r_table_add_column (t, n, "addr", 0);
+		r_table_add_column (t, n, "value", 0);
+		r_table_add_column (t, s, "refs", 0);
+	}
+	if (mode == 'j') {
+		pj = pj_new ();
+		if (!pj) {
+			return;
+		}
+	}
+	if (mode == 'j' || mode == ',' || mode == '*' || mode == 'q') {
+		size_t i;
+		const int base = core->anal->bits;
+		if (pj) {
+			pj_a (pj);
+		}
+		const ut8 *buf = core->block;
+		int withref = 0;
+		for (i = 0; i < core->blocksize; i += wordsize) {
+			ut64 addr = core->offset + i;
+			ut64 *foo = (ut64 *) (buf + i);
+			ut64 val = *foo;
+			if (base == 32) {
+				val &= UT32_MAX;
+			}
+			if (pj) {
+				pj_o (pj);
+				pj_kn (pj, "addr", addr);
+				pj_kn (pj, "value", val);
+			}
+
+			// XXX: this only works in little endian
+			withref = 0;
+			char *refs = NULL;
+			if (core->print->hasrefs) {
+				char *rstr = core->print->hasrefs (core->print->user, val, true);
+				if (rstr && *rstr) {
+					r_str_trim (rstr);
+					if (pj) {
+						char *ns = r_str_escape (rstr);
+						pj_ks (pj, "ref", r_str_trim_head_ro (ns));
+						pj_end (pj);
+						free (ns);
+					}
+					withref = 1;
+				}
+				refs = rstr;
+			}
+			if (mode == '*' && refs && *refs) {
+				// Show only the mapped ones?
+				char *rs = r_str_escape_r2 (refs);
+				r_cons_printf ("f pxr.%"PFMT64x"=0x%"PFMT64x" # %s\n", val, addr, rs);
+				free (rs);
+			}
+			if (mode == 'q' && refs && *refs) {
+				r_cons_printf ("%s\n", refs);
+			}
+			if (t) {
+				r_table_add_rowf (t, "xxs", addr, val, refs);
+				refs = NULL;
```

Comment:
doesn't `refs` leak this way?

Reply:
nope, rowf owns the pointer passed

Reply:
I don't think so. `r_table_add_rowf` does a strcpy AFAIS.

Reply:
indeed. dunno what was thinking